larq_zoo/core/utils.py

```diff
@@ -103,6 +103,41 @@ def get_input_layer(input_shape, input_tensor):
     return input_tensor
 
 
+def global_pool(x: tf.Tensor, data_format: str = "channels_last") -> tf.Tensor:
+    """Global average 2D pooling and flattening.
+
+    Alternative to existing keras implementation of GlobalPooling2D.
+    AveragePooling2D is much faster than GlobalPooling2D on Larq Compute Engine.
+
+    # Arguments
+    x: 4D TensorFlow tensor.
+    data_format: data_format: A string, one of channels_last (default) or
+        channels_first. The ordering of the dimensions in the inputs. channels_last
+        corresponds to inputs with shape (batch, height, width, channels) while
+        channels_first corresponds to inputs with shape (batch, channels, height,
+        width). It defaults to "channels_last".
+
+    # Returns
+    2D TensorFlow tensor.
+
+    # Raises
+    ValueError: if tensor is not 4D or data_format is not recognized.
+    """
+    if len(x.get_shape()) != 4:
+        raise ValueError("Tensor is not 4D.")
+    if data_format not in ["channels_last", "channels_first"]:
+        raise ValueError("data_format not recognized.")
+
+    input_shape = x.get_shape().as_list()
```

Comment:
We could in theory check here if the height and width have known, static shape, and if so use the `AveragePooling2D`, otherwise use `GlobalAveragePooling2D` if one of them is `None`.

Reply:
Good point.. I added a fix, but not sure if we want to merge this as it's getting rather complicated. @Tombana @lgeiger do you know how easy it would be to fix this at the converter?

Reply:
So *in theory* tflite has support for resizing the input tensor at runtime (i.e. resize from one inference pass to the next). This would not work well with the averagepool operator. However, since MobileNet also uses averagepool, I'm assuming they expect people to not resize their network at runtime. Therefore I think we can safely add such a step to the converter.

For reference:
The layer `GlobalAveragePooling2D` gets transformed to the tflite `Mean` operator which for float32 only has a reference implementation. They do seem to have [an optimized implementation](https://github.com/tensorflow/tensorflow/blob/5baff53e4e4507e4ae72ad2b1fae6bf750e5d808/tensorflow/lite/kernels/reduce.cc#L291) for int8 though!
The layer `AveragePooling2D` gets transformed to `AveragePool2D`, which has [an optimized implementation](https://github.com/tensorflow/tensorflow/blob/5baff53e4e4507e4ae72ad2b1fae6bf750e5d808/tensorflow/lite/kernels/pooling.cc#L113) for float.

Reply:
I think the issue is that this really should be something TFLite does automatically, it's not in the scope of LCE.

Reply:
I agree with @AdamHillier . However, for now we want a fix, and I think a temporary fix in the MLIR converter is cleaner than a temporary fix in Larq. (Assuming we could get it working properly in MLIR, that is)

Reply:
Yes if we could do this in the converter it would be great! Will leave the PR open for a bit in case unexpected difficulties come up.

Reply:
I really don't think this has a place in the converter. It should either be handled by TensorFlow, or in user space.

@koenhelwegen did you try to get the shape via `tf.shape(x)` which should get the dynamic shape of the tensor? Also this is only a problem for us in zoo since we explicitely allow for `(None, None, 3)` inputs to support different input sizes.

Reply:
`pool_size` only accepts ints, so `AveragePooling2D` won't work for models with dynamic input sizes.

I think for now the best solution is to put it here. I've simplified the implementation a bit and added a (minimalistic) unit test. Would be happy to merge this now.